integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/ObservationFhirResourceProviderIntegrationTest.java

```diff
@@ -546,4 +548,483 @@ public void shouldReturnCountForObservationAsXml() throws Exception {
+		assertThat(entries, everyItem(hasResource(
+		    hasProperty("code", hasProperty("coding", everyItem(hasProperty("display", equalTo("Respiratory rate"))))))));
+		
+		assertThat(checkDistinctObsDateTimeInEachCodeGroupIsSortedAndWithinMax(entries, 2), is(true));
+	}
+	
+	@Test
+	public void shouldReturnLastnObservationsWhenCodeIsMissingAsXml() throws Exception {
+		MockHttpServletResponse response = get(
+		    "Observation/$lastn?max=2&subject=" + OBS_PATIENT_UUID + "&category=laboratory").accept(FhirMediaTypes.XML).go();
+		
+		assertThat(response, isOk());
+		assertThat(response.getContentType(), is(FhirMediaTypes.XML.toString()));
+		assertThat(response.getContentAsString(), notNullValue());
+		
+		Bundle results = readBundleResponse(response);
+		
+		assertThat(results, notNullValue());
+		assertThat(results.getType(), equalTo(Bundle.BundleType.COLLECTION));
+		assertThat(results.hasEntry(), is(true));
+		assertThat(results, hasProperty("total", equalTo(17)));
+		
+		List<Bundle.BundleEntryComponent> entries = results.getEntry();
+		
+		assertThat(entries, everyItem(hasProperty("fullUrl", startsWith("http://localhost/ws/fhir2/R3/Observation/"))));
+		assertThat(entries, everyItem(hasResource(instanceOf(Observation.class))));
+		assertThat(entries, everyItem(hasResource(validResource())));
+		assertThat(entries,
+		    everyItem(hasResource(hasProperty("subject", hasProperty("reference", endsWith(OBS_PATIENT_UUID))))));
+		assertThat(entries, everyItem(hasResource(hasProperty("category",
+		    everyItem(hasProperty("coding", everyItem(hasProperty("code", equalTo("laboratory")))))))));
+		
+		assertThat(checkDistinctObsDateTimeInEachCodeGroupIsSortedAndWithinMax(entries, 2), is(true));
+	}
+	
+	@Test
+	public void shouldReturnLastnObservationsWhenNoParamterIsGivenAsXml() throws Exception {
+		MockHttpServletResponse response = get("Observation/$lastn?").accept(FhirMediaTypes.XML).go();
+		
+		assertThat(response, isOk());
+		assertThat(response.getContentType(), is(FhirMediaTypes.XML.toString()));
+		assertThat(response.getContentAsString(), notNullValue());
+		
+		Bundle results = readBundleResponse(response);
+		
+		assertThat(results, notNullValue());
+		assertThat(results.getType(), equalTo(Bundle.BundleType.COLLECTION));
+		assertThat(results.hasEntry(), is(true));
+		assertThat(results, hasProperty("total", equalTo(14)));
+		
+		List<Bundle.BundleEntryComponent> entries = results.getEntry();
+		
+		assertThat(entries, everyItem(hasProperty("fullUrl", startsWith("http://localhost/ws/fhir2/R3/Observation/"))));
+		assertThat(entries, everyItem(hasResource(instanceOf(Observation.class))));
+		assertThat(entries, everyItem(hasResource(validResource())));
+		
+		assertThat(checkDistinctObsDateTimeInEachCodeGroupIsSortedAndWithinMax(entries, 1), is(true));
+	}
+	
+	private boolean checkDistinctObsDateTimeInEachCodeGroupIsSortedAndWithinMax(List<Bundle.BundleEntryComponent> entries, int max) {
```

Comment:
Since this logic is being exactly replicated for R3 and R4, I think it is better to add it in a Utils class.
@ibacher Any comments/thoughts on this?